File: ldf/R/ldf.r
argmax <- function(x) {
  which.max(x)
}


first_layer <- function(agent_loglikelihoods, alphas, is_argmax, cs) {
  
  T = dim(agent_loglikelihoods)[2]
  n_agents = dim(agent_loglikelihoods)[1]
  pi_t = matrix(1, length(alphas), n_agents) / n_agents
  prob = array(0, dim=c(T, length(alphas), n_agents))
  loglikehood = matrix(0, length(alphas), T)
  
  for (t in 1:T) {
    for (j in 1:length(alphas)) {
      alpha = alphas[j]
      if (is_argmax) {
        pi_t[j, ] = alpha * pi_t[j, ]
        p = rep(0, length(pi_t[j, ]))
        p[argmax(pi_t[j, ])] = 1
        prob[t, j, ] = p
        loglikehood[j, t] = sum(agent_loglikelihoods[ , t] * p)
        pi_t[j, ] = pi_t[j, ] + agent_loglikelihoods[ , t]
      }
      else {
        pi_t[j, ] = (pi_t[j, ]^alpha + cs) / (sum(pi_t[j, ]^alpha) + cs)  # pi_t|t-1
        prob[t, j, ] = pi_t[j, ]
        loglikehood[j, t] = log(sum(exp(agent_loglikelihoods[ , t]) * pi_t[j, ]))
        pi_t[j, ] = pi_t[j, ] * exp(agent_loglikelihoods[ , t])
        pi_t[j, ] = pi_t[j, ] / sum(pi_t[j])  # pi_t|t
      }
    }
  }
  return(list("loglikehood"=loglikehood, "prob"=prob))
}

second_layer <- function(loglikehood, deltas, is_argmax) {
  T = dim(loglikehood)[2]
  res_weights = array(0, dim=c(length(deltas), T, dim(loglikehood)[1]))
  res_likelihoods = matrix(0, length(deltas), T)
  
  for (i in 1:length(deltas)) {
    delta = deltas[i]
    weights = matrix(0, T, dim(loglikehood)[1])
    momentum = rep(0,T)
    csum = matrix(0, T, dim(loglikehood)[1])
    for (t in 1:T) {
      if (is_argmax) {
        p = rep(0, dim(loglikehood)[1])
        if (t==1) {
          p[argmax(csum[t, ])] = 1
        }
        else {
          p[argmax(csum[t-1, ])] = 1
        }
        weights[t, ] = p
        momentum[t] = log(sum(exp(loglikehood[ , t]) * p))
      }
      else {
        if (t==1) {
          m = 0
          p = exp(csum[t, ] - log(sum(exp(csum[t, ] - m))) - m)
        }
        else {
          m = max(csum[t-1, ])
          p = exp(csum[t-1, ] - log(sum(exp(csum[t-1, ] - m))) - m)
        }
        weights[t, ] = p
        momentum[t] = log(sum(exp(loglikehood[ , t]) * p))
      }
      if (t==1) {
        csum[t, ] = loglikehood[ , t]
      }
      else {
        csum[t, ] = delta * csum[t-1, ] + loglikehood[ , t]
      }
    }
    res_likelihoods[i, ] = momentum
    res_weights[i, , ] = weights
  }
  return(list("res_likelihoods"=res_likelihoods, "res_weights"=res_weights))
}

ldf <- function(agent_loglikelihoods, levels, discount_factors, activation_functions, c=10^(-20)) {
  
  alphas = discount_factors
  is_argmax = activation_functions[1] == "argmax"
  res = first_layer(agent_loglikelihoods, alphas, is_argmax, c)
  loglikehood = res$loglikehood
  prob = res$prob
  
  if (levels == 2) {
    deltas = discount_factors
    is_argmax = activation_functions[2] == "argmax"
    res = second_layer(loglikehood, deltas, is_argmax)
    momentum = res$res_likelihoods
    param_weights = res$res_weights
    weights = array(0, dim=c(length(deltas), dim(forecasts_lik)[2], dim(forecasts_lik)[1]))
    for (k in 1:length(deltas)) {
      for (i in 1:dim(param_weights)[2]) {
        weights[k, i, ] = apply(t(prob[i, , ]) * param_weights[k, i, ],1,sum)
      }
    }
    list("logscores"=momentum , "weights"=weights, "params_weights"=param_weights)
  }
  else{
    list("logscores"=loglikehood , "weights"=prob, "params_weights"=NULL)
  }
  
}


forecasts_lik = t(matrix(c(
  c(-2.302585093, -4.605170186, -4.605170186, -1.609437912, -4.605170186, -1.771956842, -1.609437912, -1.386294361,
   -1.386294361, -2.302585093, -1.386294361),
  c(-0.916290732, -1.203972804, -1.609437912, -2.995732274, -4.605170186, -2.302585093, -4.828313737, -1.386294361,
   -1.609437912, -2.991548932, -0.801038303),
  c(-2.302585093, -3.688879454, -3.688879454, -1.897119985, -2.9, -2.302585093, -4.1, -1.1, -0.1, -1.4,
   -1.897119985),
  c(-4.605170186, -4.605170186, -4.605170186, -4.605170186, -4.605170186, -0.9, -4.605170186, -0.510825624,
   -0.287682072, -1.491654877, -1.897119985),
  c(-1.203972804, -1.203972804, -1.290984181, -2.302585093, -4.605170186, -1, -4.605170186, -1.049822124,
   -2.302585093, -2.995732274, -0.597837001)), nrow=11, ncol=5))
discount_factors = c(1, 0.5)

res = ldf(forecasts_lik, 1, discount_factors, c("softmax"))
stopifnot(max(abs(apply(res$logscores,1,sum) - c(-24.54868304, -21.48288988))) < 10^(-8))
res = ldf(forecasts_lik, 1, discount_factors, c("argmax"))
stopifnot(max(abs(apply(res$logscores,1,sum) - c(-24.57489777, -25.00121507))) < 10^(-8))
res = ldf(forecasts_lik, 2, discount_factors, c("softmax", "softmax"))
stopifnot(max(abs(apply(res$logscores,1,sum) - c(-22.13047414, -22.23379625))) < 10^(-8))
res = ldf(forecasts_lik, 2, discount_factors, c("argmax", "argmax"))
stopifnot(max(abs(apply(res$logscores,1,sum) - c(-25.87418075, - 24.27844848))) < 10^(-8))

File: ldf/ldf.py
import numpy as np
from typing import List
import numba as nb


@nb.jit(nopython=True, fastmath=True)
def _first_layer(agent_loglikelihoods: np.ndarray, alphas: np.ndarray, is_argmax, c):
    T = agent_loglikelihoods.shape[1]
    n_agents = int(agent_loglikelihoods.shape[0])
    pi_t = np.ones((len(alphas), n_agents)) / n_agents
    prob = np.zeros((T, len(alphas), n_agents))
    loglikehood = np.zeros((len(alphas), T))

    for t in range(T):
        for j, alpha in enumerate(alphas):
            if is_argmax:
                pi_t[j] = alpha * pi_t[j]
                p = np.zeros_like(pi_t[j])
                p[pi_t[j].argmax()] = 1
                prob[t, j] = p.copy()
                loglikehood[j, t] = (agent_loglikelihoods[:, t] * p).sum()
                pi_t[j] = pi_t[j] + agent_loglikelihoods[:, t]
            else:
                pi_t[j] = (pi_t[j] ** alpha + c) / (np.sum(pi_t[j] ** alpha) + c)  # pi_t|t-1
                prob[t, j] = pi_t[j].copy()
                loglikehood[j, t] = np.log((np.exp(agent_loglikelihoods[:, t]) * pi_t[j]).sum())
                pi_t[j] = pi_t[j] * np.exp(agent_loglikelihoods[:, t])
                pi_t[j] = pi_t[j] / np.sum(pi_t[j])  # pi_t|t

    return loglikehood, prob


@nb.jit(nopython=True, fastmath=True)
def _second_layer(loglikehood: np.ndarray, deltas: np.ndarray, is_argmax, c):

    T = loglikehood.shape[1]
    res_weights = np.empty((len(deltas), T, loglikehood.shape[0]))
    res_likelihoods = np.empty((len(deltas), T))

    for i, delta in enumerate(deltas):
        weights = np.zeros((T, loglikehood.shape[0]))
        momentum = np.empty(T)
        cumsum = np.zeros((T, loglikehood.shape[0]))

        for t in range(T):
            if is_argmax:
                p = np.zeros(loglikehood.shape[0])
                p[np.argmax(cumsum[t-1])] = 1
                weights[t] = p
                momentum[t] = np.log((np.exp(loglikehood[:, t]) * p).sum())
            else:
                m = max(cumsum[t-1])
                p = np.exp(cumsum[t-1] - np.log(np.sum(np.exp(cumsum[t-1] - m))) - m)
                weights[t] = p
                momentum[t] = np.log((np.exp(loglikehood[:, t]) * p).sum())
            cumsum[t] = delta * cumsum[t-1] + loglikehood[:, t]

        res_likelihoods[i] = momentum
        res_weights[i] = weights

    return res_likelihoods, res_weights


def ldf(agent_loglikelihoods: np.ndarray, levels, discount_factors: List, activation_functions=None, c=10**-20):
    """

    :param agent_loglikelihoods:
    :param levels:
    :param discount_factors:
    :param activation_functions:
    :param c:
    :return:
    """
    alphas = discount_factors[0]
    assert len(discount_factors) == levels
    if not (isinstance(alphas, list) or isinstance(alphas, np.ndarray)):
        alphas = np.array([alphas])

    is_argmax = activation_functions[0] == "argmax"
    loglikehood ,prob = _first_layer(agent_loglikelihoods, alphas, is_argmax, c)
    prob = prob.transpose(1, 0, 2)

    if levels == 1:
        return {"logscores": loglikehood, "weights": prob, "params_weights": None}
    else:
        for k in range(levels-1):
            deltas = discount_factors[k+1]
            if not (isinstance(deltas, list) or isinstance(deltas, np.ndarray)):
                deltas = np.array([deltas])

            is_argmax = activation_functions[k+1] == "argmax"
            loglikehood, param_weights = _second_layer(loglikehood, deltas, is_argmax, c)
            weights = []

            for k in range(len(deltas)):  # iteration over discount factors
                weights.append(
                    np.asarray([(prob[:, i, :].T * j).T.sum(axis=0) for i, j in enumerate(param_weights[k])])) # iteration over T
            prob = np.asarray(weights)
        return {"logscores": loglikehood, "weights": prob, "params_weights": param_weights}